packages/stellar-axelar-std-derive/src/storage.rs

```diff
@@ -0,0 +1,176 @@
+use heck::ToSnakeCase;
+use proc_macro2::{Ident, TokenStream};
+use quote::{format_ident, quote, ToTokens};
+use syn::{Attribute, Data, DataEnum, DeriveInput, Fields, FieldsNamed, Meta, Type, Variant};
+
+#[derive(Default)]
+struct StorageAttributes {
+    storage_type: StorageType,
+    value_type: Option<Type>,
+}
+
+#[derive(Default, Debug)]
+enum StorageType {
+    #[default]
+    Instance,
+    Persistent,
+    Temporary,
+}
+
+pub fn contractstorage(input: &DeriveInput) -> TokenStream {
+    let name = &input.ident;
+
+    let Data::Enum(DataEnum { variants, .. }) = &input.data else {
+        panic!("contractstorage can only be used on enums.");
+    };
+
+    let transformed_variants: Vec<_> = variants.iter().map(transform_variant).collect();
+
+    let storage_fns: Vec<_> = variants
+        .iter()
+        .map(|variant| {
+            let attrs = storage_attributes(&variant.attrs);
+            storage_fns(name, variant, &attrs)
+        })
+        .collect();
+
+    let output = quote! {
+        #[contracttype]
+        pub enum #name {
+            #(#transformed_variants,)*
+        }
+
+        impl #name {
+            #(#storage_fns)*
+        }
+    };
+
+    output
+}
+
+fn transform_variant(variant: &Variant) -> TokenStream {
+    let variant_name = &variant.ident;
+
+    match &variant.fields {
+        Fields::Unit => {
+            quote! {
+                #variant_name
+            }
+        }
+        Fields::Named(FieldsNamed { named, .. }) => {
+            let types = named.iter().map(|f| &f.ty);
+            quote! {
+                #variant_name(#(#types),*)
+            }
+        }
+        _ => panic!("Only unit variants or named fields are supported in storage enums."),
+    }
+}
+
+fn storage_attributes(attrs: &[Attribute]) -> StorageAttributes {
+    let mut storage_attributes = StorageAttributes::default();
```

Comment:
bad pattern to use mutation. separate this into two functions, one to return the storage type, and another for the value. Ensure that exactly one distinct attribute is present for storage type, and one for the value